### ISSUE TITLE
Spike: Can we switch to `containerd` mode in Garden directly?

### ISSUE THREAD

**Issue by ddadlani** [concourse/concourse#4130]
We plan to switch over to `containerd` eventually (#3891) but as a first step we might just be able to run garden in `containerd` mode (instead of `runc`). The API should remain the same for us, so let's see if we can just switch it and run testflight/topgun with no issues.

If there are issues, let's document them here for further investigation.

**Comment by kcmannem** [comment]
take a look at https://github.com/concourse/concourse/issues/4169

**Comment by deniseyu** [comment]
Update of spike:

We managed to get `containerd` running within Garden, using docker-compose locally!

Here's what we did:

1. Add the runtime flags to toggle Garden over to using containerd in `cmd/concourse/worker_linux.go`:
  ```
	gdnServerFlags := []string{
		"--bind-ip", cmd.BindIP.IP.String(),
		"--bind-port", fmt.Sprintf("%d", cmd.BindPort),

		"--depot", depotDir,
		"--properties-path", filepath.Join(cmd.WorkDir.Path(), "garden-properties.json"),

		"--time-format", "rfc3339",

		// disable graph and grootfs setup; all images passed to Concourse
		// containers are raw://
		"--no-image-plugin",

          // new flags for containerd
		"--use-containerd-for-processes", "true",
		"--containerd-socket", "/run/containerd/containerd.sock",
	}
  ```
2. While configuring this we noticed that `garden` does not bring up the `containerd` process. BOSH does this for us via monit because `containerd` ships as a job inside of the garden release. So we have to kick off the containerd process ourselves. We used a new Dockerfile to run the workers that builds on top of the web image:
  ```
FROM concourse/concourse-web:local AS dev
# don't need to do anything here, but we need to preload this image so we can copy keys out
FROM ubuntu:disco AS ubuntu
FROM ubuntu

RUN mkdir -p /concourse-keys
COPY --from=dev /concourse-keys/ /concourse-keys/

# 'web' keys
ENV CONCOURSE_SESSION_SIGNING_KEY     /concourse-keys/session_signing_key
ENV CONCOURSE_TSA_AUTHORIZED_KEYS     /concourse-keys/authorized_worker_keys
ENV CONCOURSE_TSA_HOST_KEY            /concourse-keys/tsa_host_key

# 'worker' keys
ENV CONCOURSE_TSA_PUBLIC_KEY          /concourse-keys/tsa_host_key.pub
ENV CONCOURSE_TSA_WORKER_PRIVATE_KEY  /concourse-keys/worker_key

# enable DNS proxy to support Docker's 127.x.x.x DNS server
ENV CONCOURSE_GARDEN_DNS_PROXY_ENABLE         true
ENV CONCOURSE_WORKER_GARDEN_DNS_PROXY_ENABLE  true

# auto-wire work dir for 'worker' and 'quickstart'
ENV CONCOURSE_WORK_DIR                /worker-state
ENV CONCOURSE_WORKER_WORK_DIR         /worker-state

# volume for non-aufs/etc. mount for baggageclaim's driver
VOLUME /worker-state

RUN apt update && apt install -y \
    btrfs-tools \
    ca-certificates \
    dumb-init \
    iproute2 \
    libc6 \
    wget \
    tar \
    file

COPY --from=dev /usr/local/concourse /usr/local/concourse
STOPSIGNAL SIGUSR2

# get runc binary
RUN wget https://github.com/opencontainers/runc/releases/download/v1.0.0-rc8/runc.amd64
RUN chmod +x runc.amd64
RUN mv runc.amd64 /usr/local/bin/runc

# start containerd manually
RUN wget https://github.com/containerd/containerd/releases/download/v1.3.0-beta.0/containerd-1.3.0-beta.0.linux-amd64.tar.gz
RUN tar -zxf containerd-1.3.0-beta.0.linux-amd64.tar.gz

COPY foo .

ENTRYPOINT ["/usr/bin/bash"]
  ```
where `foo` is a script that backgrounds containerd and runs the Concourse binary (there is probably a more elegant way to do this, but meh, it's a spike)
  ```
#!/bin/bash
/bin/containerd &
/usr/local/concourse/bin/concourse worker
  ```
3. The worker node started up now, but errored out because `containerd` looks for a `runc` binary in the path. We added the section to pull the runc binary, restarted, set a pipeline, and it Just Worked ™️ 

### Next steps

- [ ] Run testflight
- [ ] Package for bosh-lite
- [ ] Run Topgun tests
- [ ] Remove the process hacking from Dockerfiles and get the `containerd` process management logic into `cmd/concourse/worker.go`
- [ ] Think about we want to build `containerd` and `runc`, or package the binaries

### Other things to think about later:

* **Linux kernel discrepancy**: We build `concourse/dev` using an official `golang` image, which uses `debian` underneath. We cannot run `containerd` in Debian because it depends on the `seccomp_api_set` filesystem capability to be present in the kernel API which Debian doesn't have but Ubuntu does. We also can't just bump our dev image to use Bionic as the base layer because our Concourse binary requires `glibc` version 2.28 or higher, which first appears in Cosmic! For this spike we used Cosmic and Disco to compile Concourse. 
The version of concourse/dev was `6cb525d4090` 4.48GB
  ```
/usr/local/concourse/bin/concourse: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /usr/local/concourse/bin/concourse)
  ```
* `containerd` can be declaratively configured using a `containerd.toml` file. We used the defaults for this spike and they worked fine, but we may want to tune them one day. Here is an example from the garden-runc-release: https://github.com/cloudfoundry/garden-runc-release/blob/develop/jobs/garden/templates/config/containerd.toml.erb
* Licensing for containerd

cc @vito @concourse/runtime

**Comment by deniseyu** [comment]
Update: We were able to successfully run the testflight suite

**Comment by ddadlani** [comment]
Our next step is to create a bosh release for `garden-containerd` and run topgun as a sanity check.

**Comment by ddadlani** [comment]
@xtreme-sameer-vohra and @deniseyu ran into issues with running containerd on Bionic because of a dependency on `glibc`. As a result this spike has been conducted on Disco/Cosmic. We may need to statically link `glibc` to avoid causing issues for our users.

**Comment by xtreme-sameer-vohra** [comment]
For Topgun tests
@kcmannem & I were able to build a bosh-release with containerd and got that deployed on BOSH such that Concourse binary started successfully.
However, we ran into where when executing a build, Concourse complained about the `runc` binary not being available.

**Comment by xtreme-sameer-vohra** [comment]
The next steps are to modify the bosh-release to be able to run the complete the Topgun suite

**Comment by kcmannem** [comment]
we used the default containerd config toml to spin up containerd. The problem with this is that the runc binay cannot be found. (for some reason the path variable is empty).
```
  [plugins.linux]
    shim = "containerd-shim"
    runtime = "runc"
```
The runtime key should point to the absolute path of the runc binary if its not found in the PATH variable.
We fixed the bosh release by changing the above portion of the toml to
```
  [plugins.linux]
    shim = "containerd-shim"
    runtime = "/usr/local/bin/runc"
```
